### PR TITLE
Invanti Security Controls - Updated Start Scan Action

### PR DESCRIPTION
**Pull request by dmandyna-r7**
## Proposed Changes

Added creation of session credentials and changed polling method

### Description

Describe the proposed changes:

  - Creating session credentials before the action
  - Polling status using Operation-Location header

### Testing

Any testing information goes here.

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Style

Review the [style guide](https://komand.github.io/python/style.html)

- [ ] For dependencies, pin [OS package](https://komand.github.io/python/style.html#dockerfile) and [Python package](https://komand.github.io/python/style.html#requirements-txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://komand.github.io/python/sdk.html#sdk-versions) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://komand.github.io/python/error_handling.html#plugin-exceptions) and [ConnectionTestException](https://komand.github.io/python/error_handling.html#connection-exceptions)
- [ ] For logging, use [self.logger](https://komand.github.io/python/sdk.html#logging)
- [ ] For docs, use [changelog style](https://komand.github.io/python/style.html#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://komand.github.io/python/style.html#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

#### Run

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Ivanti Security Controls:1.2.1. Step name: start_patch_scan\n",
    "meta": {},
    "output": {
      "scan_details": {
        "id": "724f9202-a0de-44f5-99be-5ef7a4d877e9",
        "isComplete": true,
        "links": {
          "self": {
            "href": "https://ivanti-w16.vuln.lax.rapid7.com:3121/st/console/api/v1.0/patch/scans/724f9202-a0de-44f5-99be-5ef7a4d877e9"
          }
        },
        "name": "test-scan-1",
        "scanType": "Patch",
        "startedOn": "2020-06-28T15:48:01.8258245Z",
        "updatedOn": "2020-06-28T15:50:23.877Z",
        "user": "IVANTI-W16\\Administrator"
      }
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/ivanti_security_controls:1.2.1 --debug run < tests/start_patch_scan_1.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Ivanti Security Controls:1.2.1. Step name: start_patch_scan\n",
    "meta": {},
    "output": {
      "scan_details": {
        "id": "22b589aa-8ca6-4d43-9f30-124aff733a2f",
        "isComplete": true,
        "links": {
          "self": {
            "href": "https://ivanti-w16.vuln.lax.rapid7.com:3121/st/console/api/v1.0/patch/scans/22b589aa-8ca6-4d43-9f30-124aff733a2f"
          }
        },
        "name": "test-scan-2",
        "scanType": "Patch",
        "startedOn": "2020-06-28T15:50:52.3339905Z",
        "updatedOn": "2020-06-28T15:51:17.68Z",
        "user": "IVANTI-W16\\Administrator"
      }
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/ivanti_security_controls:1.2.1 --debug run < tests/start_patch_scan_2.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Ivanti Security Controls:1.2.1. Step name: start_patch_scan\n",
    "meta": {},
    "output": {
      "scan_details": {
        "id": "1f135070-a03f-43d8-ac60-3ab76d0293c8",
        "isComplete": true,
        "links": {
          "self": {
            "href": "https://ivanti-w16.vuln.lax.rapid7.com:3121/st/console/api/v1.0/patch/scans/1f135070-a03f-43d8-ac60-3ab76d0293c8"
          }
        },
        "name": "test-scan-3",
        "scanType": "Patch",
        "startedOn": "2020-06-28T15:51:48.2937223Z",
        "updatedOn": "2020-06-28T15:54:39.68Z",
        "user": "IVANTI-W16\\Administrator"
      }
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/ivanti_security_controls:1.2.1 --debug run < tests/start_patch_scan_3.json
</summary>
</details>

<details>

```
{
  "body": {
    "error": "An error occurred during plugin execution!\n\nInvalid username or password provided. Verify your username and password are correct.",
    "log": "Connect: Connecting...\nrapid7/Ivanti Security Controls:1.2.1. Step name: start_patch_scan\nAn error occurred during plugin execution!\n\nInvalid username or password provided. Verify your username and password are correct.\nTraceback (most recent call last):\n  File \"/usr/local/lib/python3.8/site-packages/insightconnect_plugin_runtime-4.0.2-py3.8.egg/insightconnect_plugin_runtime/plugin.py\", line 307, in handle_step\n    output = self.start_step(\n  File \"/usr/local/lib/python3.8/site-packages/insightconnect_plugin_runtime-4.0.2-py3.8.egg/insightconnect_plugin_runtime/plugin.py\", line 428, in start_step\n    output = func(params)\n  File \"/usr/local/lib/python3.8/site-packages/ivanti_security_controls_rapid7_plugin-1.2.1-py3.8.egg/icon_ivanti_security_controls/actions/start_patch_scan/action.py\", line 42, in run\n    self.connection.ivanti_api.create_session_credential()\n  File \"/usr/local/lib/python3.8/site-packages/ivanti_security_controls_rapid7_plugin-1.2.1-py3.8.egg/icon_ivanti_security_controls/util/api.py\", line 117, in create_session_credential\n    self._call_api(\"DELETE\", f\"{self.url}/sessioncredentials\", allow_404=True, return_response=True)\n  File \"/usr/local/lib/python3.8/site-packages/ivanti_security_controls_rapid7_plugin-1.2.1-py3.8.egg/icon_ivanti_security_controls/util/api.py\", line 131, in _call_api\n    raise PluginException(preset=PluginException.Preset.USERNAME_PASSWORD)\ninsightconnect_plugin_runtime.exceptions.PluginException: An error occurred during plugin execution!\n\nInvalid username or password provided. Verify your username and password are correct.\n",
    "meta": {},
    "status": "error"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/ivanti_security_controls:1.2.1 --debug run < tests/start_patch_scan_invalid_creds_bad.json
</summary>
</details>

<details>

```
{
  "body": {
    "error": "An error occurred during plugin execution!\n\nSomething unexpected occurred. Check the logs and if the issue persists please contact support. Response was: {\"message\":\"The request is invalid.\",\"modelState\":{\"scanInput.templateId\":[\"An error has occurred.\",\"TemplateId must be specified.\"]}}",
    "log": "Connect: Connecting...\nrapid7/Ivanti Security Controls:1.2.1. Step name: start_patch_scan\nAn error occurred during plugin execution!\n\nSomething unexpected occurred. Check the logs and if the issue persists please contact support. Response was: {\"message\":\"The request is invalid.\",\"modelState\":{\"scanInput.templateId\":[\"An error has occurred.\",\"TemplateId must be specified.\"]}}\nTraceback (most recent call last):\n  File \"/usr/local/lib/python3.8/site-packages/insightconnect_plugin_runtime-4.0.2-py3.8.egg/insightconnect_plugin_runtime/plugin.py\", line 307, in handle_step\n    output = self.start_step(\n  File \"/usr/local/lib/python3.8/site-packages/insightconnect_plugin_runtime-4.0.2-py3.8.egg/insightconnect_plugin_runtime/plugin.py\", line 428, in start_step\n    output = func(params)\n  File \"/usr/local/lib/python3.8/site-packages/ivanti_security_controls_rapid7_plugin-1.2.1-py3.8.egg/icon_ivanti_security_controls/actions/start_patch_scan/action.py\", line 44, in run\n    scan = self.connection.ivanti_api.start_patch_scan(payload)\n  File \"/usr/local/lib/python3.8/site-packages/ivanti_security_controls_rapid7_plugin-1.2.1-py3.8.egg/icon_ivanti_security_controls/util/api.py\", line 93, in start_patch_scan\n    return self._call_api(\"POST\", f\"{self.url}/patch/scans\", json_data=payload, return_response=True)\n  File \"/usr/local/lib/python3.8/site-packages/ivanti_security_controls_rapid7_plugin-1.2.1-py3.8.egg/icon_ivanti_security_controls/util/api.py\", line 142, in _call_api\n    raise PluginException(preset=PluginException.Preset.UNKNOWN, data=response.text)\ninsightconnect_plugin_runtime.exceptions.PluginException: An error occurred during plugin execution!\n\nSomething unexpected occurred. Check the logs and if the issue persists please contact support. Response was: {\"message\":\"The request is invalid.\",\"modelState\":{\"scanInput.templateId\":[\"An error has occurred.\",\"TemplateId must be specified.\"]}}\n",
    "meta": {},
    "status": "error"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/ivanti_security_controls:1.2.1 --debug run < tests/start_patch_scan_invalid_template_id_bad.json
</summary>
</details>

<details>

```
{
  "body": {
    "error": "An error occurred during plugin execution!\n\nNo hostnames or machine group IDs specified. Either hostnames or machine group IDs must be specified.",
    "log": "Connect: Connecting...\nrapid7/Ivanti Security Controls:1.2.1. Step name: start_patch_scan\nAn error occurred during plugin execution!\n\nNo hostnames or machine group IDs specified. Either hostnames or machine group IDs must be specified.\nTraceback (most recent call last):\n  File \"/usr/local/lib/python3.8/site-packages/insightconnect_plugin_runtime-4.0.2-py3.8.egg/insightconnect_plugin_runtime/plugin.py\", line 307, in handle_step\n    output = self.start_step(\n  File \"/usr/local/lib/python3.8/site-packages/insightconnect_plugin_runtime-4.0.2-py3.8.egg/insightconnect_plugin_runtime/plugin.py\", line 428, in start_step\n    output = func(params)\n  File \"/usr/local/lib/python3.8/site-packages/ivanti_security_controls_rapid7_plugin-1.2.1-py3.8.egg/icon_ivanti_security_controls/actions/start_patch_scan/action.py\", line 24, in run\n    raise PluginException(cause='No hostnames or machine group IDs specified.',\ninsightconnect_plugin_runtime.exceptions.PluginException: An error occurred during plugin execution!\n\nNo hostnames or machine group IDs specified. Either hostnames or machine group IDs must be specified.\n",
    "meta": {},
    "status": "error"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/ivanti_security_controls:1.2.1 --debug run < tests/start_patch_scan_no_host_name_bad.json
</summary>
</details>

<details>

```
{
  "body": {
    "error": "An error occurred during plugin execution!\n\nAction timeout. This scan has exceeded the maximum poll time of 15.",
    "log": "Connect: Connecting...\nrapid7/Ivanti Security Controls:1.2.1. Step name: start_patch_scan\nAn error occurred during plugin execution!\n\nAction timeout. This scan has exceeded the maximum poll time of 15.\nTraceback (most recent call last):\n  File \"/usr/local/lib/python3.8/site-packages/ivanti_security_controls_rapid7_plugin-1.2.1-py3.8.egg/icon_ivanti_security_controls/actions/start_patch_scan/action.py\", line 49, in run\n    polling2.poll(lambda: self.connection.ivanti_api.get_operation_location(operation_location_url)\n  File \"/usr/local/lib/python3.8/site-packages/polling2.py\", line 192, in poll\n    raise TimeoutException(values, last_item)\npolling2.TimeoutException: (\u003cqueue.Queue object at 0x7fb4355378b0\u003e, False)\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/usr/local/lib/python3.8/site-packages/insightconnect_plugin_runtime-4.0.2-py3.8.egg/insightconnect_plugin_runtime/plugin.py\", line 307, in handle_step\n    output = self.start_step(\n  File \"/usr/local/lib/python3.8/site-packages/insightconnect_plugin_runtime-4.0.2-py3.8.egg/insightconnect_plugin_runtime/plugin.py\", line 428, in start_step\n    output = func(params)\n  File \"/usr/local/lib/python3.8/site-packages/ivanti_security_controls_rapid7_plugin-1.2.1-py3.8.egg/icon_ivanti_security_controls/actions/start_patch_scan/action.py\", line 52, in run\n    raise PluginException(\ninsightconnect_plugin_runtime.exceptions.PluginException: An error occurred during plugin execution!\n\nAction timeout. This scan has exceeded the maximum poll time of 15.\n",
    "meta": {},
    "status": "error"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/ivanti_security_controls:1.2.1 --debug run < tests/start_patch_scan_poll_timeout_bad.json
</summary>
</details>

<details>

```
[*] Validating plugin with all validators at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator ExceptionValidator
[*] Executing validator CredentialsValidator
[*] Executing validator PasswordValidator
[*] Executing validator PrintValidator
[*] Executing validator ConfidentialValidator
[*] Executing validator DockerValidator
[*] Executing validator URLValidator
WARNING: URLs found that return a 4xx code. Verify they are publicly accessible and if not, update with a working URL.
violation: help.md[177]: https://example.com:3121/st/console/api/v1.0/patches/5033
violation: help.md[237]: https://example.com:3121/st/console/api/v1.0/patches/5033
violation: help.md[94]: https://example.com:3121/st/console/api/v1.0/patch/deploytemplates/7b5bc7e4-7437-47ac-ae2e-569ffdb0ccf8
violation: help.md[109]: https://example.com:3121/st/console/api/v1.0/patch/deployments/282cbbf9-276d-4d36-a96d-6e55c8a7271e/machines/36
violation: help.md[88]: https://example.com:3121/st/console/api/v1.0/patch/deployments/282cbbf9-276d-4d36-a96d-6e55c8a7271e
violation: help.md[91]: https://example.com:3121/st/console/api/v1.0/patch/deployments/282cbbf9-276d-4d36-a96d-6e55c8a7271e
violation: help.md[109]: https://example.com:3121/st/console/api/v1.0/patch/deployments/282cbbf9-276d-4d36-a96d-6e55c8a7271e
violation: help.md[91]: https://example.com:3121/st/console/api/v1.0/patch/deployments/282cbbf9-276d-4d36-a96d-6e55c8a7271e/machines
violation: help.md[109]: https://example.com:3121/st/console/api/v1.0/patch/deployments/282cbbf9-276d-4d36-a96d-6e55c8a7271e/machines
violation: help.md[429]: https://localhost.com:3121/st/console/api/v1.0/patch/downloads/0001f77f-0000-0000-0000-000000000000?culture=en-US
[*] Plugin successfully validated!

----
[*] Total time elapsed: 15808.624ms
```

<summary>
icon-validate --all .
</summary>
</details>

### CODE REVIEW

File: ivanti_security_controls/icon_ivanti_security_controls/actions/start_patch_scan/action.py [hunk 1]
```diff
@@ -39,19 +38,25 @@ def run(self, params={}):
...
-                    scan_details['isComplete'] = patch_scan_status_details['isComplete']
-                    scan_details['updatedOn'] = patch_scan_status_details['updatedOn']
-                    break
-            i += 10
+        
+        self.connection.ivanti_api.create_session_credential()
+
+        scan = self.connection.ivanti_api.start_patch_scan(payload)
+
+        operation_location_url = scan.headers.get("Operation-Location")
```
Comment (zyoutz-r7) [inline ivanti_security_controls/icon_ivanti_security_controls/actions/start_patch_scan/action.py:46]: Might be worth adding a check for `None` based on some of the inconsistencies and changes over time we have seen with this API.

Comment (zyoutz-r7) [inline ivanti_security_controls/icon_ivanti_security_controls/actions/start_patch_scan/action.py:46]: Might be worth adding a check for None based on some of the inconsistencies and changes over time we have seen with this API.  Or a try/except on KeyError

Comment (zyoutz-r7) [inline ivanti_security_controls/icon_ivanti_security_controls/actions/start_patch_scan/action.py:46]: Might be worth adding a check for None based on some of the inconsistencies and changes over time we have seen with this API. Or a try/except on KeyError

File: ivanti_security_controls/icon_ivanti_security_controls/actions/start_patch_scan/action.py [hunk 2]
```diff
@@ -39,19 +38,25 @@ def run(self, params={}):
...
+
+        try:
+            polling2.poll(lambda: self.connection.ivanti_api.get_operation_location(operation_location_url)
+            .get("percentComplete") == 100, step=10, timeout=max_poll_time)
```
Comment (zyoutz-r7) [inline ivanti_security_controls/icon_ivanti_security_controls/actions/start_patch_scan/action.py:50]: Log for `KeyError` if `percentComplete` isn't returned

File: ivanti_security_controls/icon_ivanti_security_controls/actions/start_patch_scan/action.py [hunk 3]
```diff
@@ -3,8 +3,7 @@
 
 # Custom imports below
 from insightconnect_plugin_runtime.exceptions import PluginException
-import time
-
+import polling2
```
Comment (zyoutz-r7) [inline ivanti_security_controls/icon_ivanti_security_controls/actions/start_patch_scan/action.py:6]: Open question for @mrinehart-r7 @jschipp-r7, do we want to include little libraries for polling, etc?  Seems like another dependency to manage/update but I get that we use it often.

File: ivanti_security_controls/icon_ivanti_security_controls/actions/start_patch_scan/action.py (hunk 2 above)
Comment (dmandyna-r7) [reply ivanti_security_controls/icon_ivanti_security_controls/actions/start_patch_scan/action.py:50]: @zyoutz-r7 I've noticed that this operation-location URL doesn't always contain percentComplete straight away. It happens after around 20-40 seconds, that's why I'm not logging this exception.

File: ivanti_security_controls/icon_ivanti_security_controls/actions/start_patch_scan/action.py (hunk 1 above)
Comment (dmandyna-r7) [reply ivanti_security_controls/icon_ivanti_security_controls/actions/start_patch_scan/action.py:46]: Thanks, that's a good idea, I'll get that added.

File: ivanti_security_controls/icon_ivanti_security_controls/actions/start_patch_scan/action.py (hunk 3 above)
Comment (jschipp-r7) [reply ivanti_security_controls/icon_ivanti_security_controls/actions/start_patch_scan/action.py:6]: This one is fine with me 👍

**Pull request merged**